src/main/scala/mesosphere/util/ZookeeperDirectState.scala

```diff
@@ -0,0 +1,186 @@
+      }
+    }
+  }
+
+  /**
+    * Marshall a variable to an Entry and store it in Zookeeper
+    */
+  private def handleStore(lv: LocalVariable): Future[Variable] = Future {
+
+    val persistedVariable = fetch(lv.name).get.asInstanceOf[LocalVariable]
+    val inputMatchesPersistedVersion = persistedVariable.storedUuid == lv.storedUuid
+
+    if (inputMatchesPersistedVersion) {
+
+      val entryData = mesosphere.marathon.Protos.Entry.newBuilder()
+        .setName(lv.name)
+        .setUuid(ByteString.copyFromUtf8(UUID.randomUUID().toString))
+        .setValue(ByteString.copyFrom(lv.value()))
+        .build()
+        .toByteArray
+
+      mkDir(Some(lv.name))
+      client.setData().forPath(fullPath(lv.name), entryData)
+      fetch(lv.name).get()
+    }
+    else {
+      log.warn(s"Attempt was made to persist ${lv.name} from obsolete version")
+      null
+    }
+
+  }
+
+  /**
+    * Remove the variable from our Zookeeper node. Returns false if deletion was not necessary.
+    */
+  private def handleExpunge(lv: LocalVariable): Future[Boolean] = Future {
+
+    if (exists(lv.name)) {
+      client.delete().forPath(fullPath(lv.name))
+      true
+    }
+    else false
+  }
+
+  /**
+    * Return the children of a Zookeeper node
+    */
+  private def handleNames: Future[util.Iterator[String]] = {
+    Future { client.getChildren.forPath(zNode).iterator() }
+  }
+
+  private def fullPath(name: String) = s"$zNode/$name"
+
+  private def mkDir(name: Option[String] = None, zkClient: CuratorFramework = client): Unit = {
+    val path = name map fullPath getOrElse zNode
+    ZKPaths.mkdirs(zkClient.getZookeeperClient.getZooKeeper, path)
+  }
+
+  private def exists(name: String): Boolean = {
+    client.checkExists().forPath(fullPath(name)) != null
+  }
+
+  /**
+    * Creates our Zookeeper client
+    */
+  private def createZookeeperClient(): Future[CuratorFramework] = {
+
+    val RETRY_POLICY = new ExponentialBackoffRetry(1000, 4)
+
+    val client = CuratorFrameworkFactory.newClient(
+      serverList,
+      sessionTimeout.toMillis.toInt,
+      connectionTimeout.toMillis.toInt,
+      RETRY_POLICY
+    )
+
+    client.start()
+
+    // Make sure we have a successful connection before returning the client
+    Future {
+      val connected = client.getZookeeperClient.blockUntilConnectedOrTimedOut()
+      if (!connected) throw new TimeoutException("Could not connect to Zookeeper")
+
+      // Create our znode, if necessary
+      mkDir(zkClient = client)
+
+      client
+    }
+  }
+
+  // Wrap our local Scala-Future-returning methods with Java-Future-returning methods
+  // to satisfy the base State interface.
+  override def names(): JFuture[util.Iterator[String]] = handleNames.asJava(sessionTimeout)
+  override def fetch(name: String): JFuture[Variable] = handleFetch(name).asJava(sessionTimeout)
+  override def store(v: Variable): JFuture[Variable] =
+    handleStore(v.asInstanceOf[LocalVariable]).asJava(sessionTimeout)
```

Comment:
Admittedly it doesn't look very good. However, this is replacing C++ code that performs the same typecast. Would it be better to remove the use of the JNI-based State and replace it with an implementation that won't require typecasting?